Search: Fairer balancing when routing searches by session ID

markharwood opened the pull request:
A user reported uneven balancing of load on nodes handling search requests from Kibana which supplies a session ID in a routing preference. Each shardId was selecting the same node for a given session ID from the list of allocations (in their case 2 data nodes with all primaries on one node and replicas on the other).
This change counteracts the tendency to opt for the same node given the same user-supplied preference by incorporating shard ID in the hash of the preference key. This will help randomise node choices across shards.

Closes #24642

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java:
```diff
@@ -177,10 +178,17 @@ private ShardIterator preferenceActiveShardIterator(IndexShardRoutingTable index
+            // Better overall balancing can be achieved if each shardId opts
+            // for a different selection from the elements in the list
+            // given the same user-supplied preference key.
+            routingHash = 31 * routingHash + indexShard.shardId.id();
```
can we use `indexShard.shardId.hashCode()` - it's good to have different hashing across indices as well.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java:
```diff
@@ -177,10 +178,17 @@ private ShardIterator preferenceActiveShardIterator(IndexShardRoutingTable index
         // if not, then use it as the index
+        int routingHash = Murmur3HashFunction.hash(preference);
+        if (nodes.getMinNodeVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
+            // Better overall balancing can be achieved if each shardId opts
```
can we say that the shard ordering is the determined by a fixed node order (and point to the AllocationService for details) and thus we need to make sure we scatter across shards? it's better to say why than just saying better ;)

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java:
```diff
@@ -367,6 +368,50 @@ public void testPreferNodes() throws InterruptedException, IOException {
+            clusterService = ClusterServiceUtils.createClusterService(threadPool);
+            final String indexName = "test";
+            final int numShards = 4;
+            ClusterServiceUtils.setState(clusterService,
```
why do we need the cluster service, can't we just use the generated state?

markharwood replied:
Copied over from testPreferNodes(). It looks like it's needed as OperationRouting.getShards(...) fails with Index not found exception without this.

markharwood replied:
Copy and pasted from other preference test.  Will remove

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java:
```diff
@@ -367,6 +368,50 @@ public void testPreferNodes() throws InterruptedException, IOException {
+            final String indexName = "test";
+            final int numShards = 4;
+            ClusterServiceUtils.setState(clusterService,
+                    ClusterStateCreationUtils.stateWithAssignedPrimariesAndOneReplica(indexName, numShards));
```
can we use random number of replicas, with more than one sometimes ?

markharwood replied:
ClusterStateCreationUtils could use an additional helper method to support this?

stateWithActivePrimary(..) lets me pick num replicas but not shards.
stateWithAssignedPrimariesAndOneReplica(...) lets me pick num shards but not replicas

bleskes replied:
sure

markharwood replied:
So with random num shards and replicas I'm going to assume for this balancing test that shard assignments are optimal i.e. num nodes = shards x replicas (or only one Lucene index on each node). Otherwise I can't reason simply about how "bunched up" replicas are on each node and the expected number of nodes servicing a search

bleskes replied:
@markharwood remember to take hash % copies# collisions into account. So if you see two shards being search on the same node, they should have the same hash modulo number of shard copies.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java:
```diff
@@ -367,6 +368,50 @@ public void testPreferNodes() throws InterruptedException, IOException {
+            threadPool = new TestThreadPool("testFairSessionIdPreferences");
+            clusterService = ClusterServiceUtils.createClusterService(threadPool);
+            final String indexName = "test";
+            final int numShards = 4;
```
randomize?

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java:
```diff
@@ -367,6 +368,50 @@ public void testPreferNodes() throws InterruptedException, IOException {
+                List<ShardRouting> searchedShards = new ArrayList<>(numShards);
+                Set<String> selectedNodes = new HashSet<>(numShards);
+                for (int shardNum = 0; shardNum < numShards; shardNum++) {
+                    final ShardIterator it = opRouting.getShards(clusterService.state(), indexName, shardNum, "my_session_id");
```
can we randomize the session id?

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java:
```diff
@@ -367,6 +368,50 @@ public void testPreferNodes() throws InterruptedException, IOException {
+                } else {
+                    assertEquals(firstSearch, searchedShards);
+                }
+                assertThat("More than one node should be selected", selectedNodes.size(), greaterThan(1));
```
can we check that the distribution is balanced in a multi index / multi shard case? like that thing differ at most with 1.

markharwood commented:
@bleskes I added randomized indexes, shards, sessions etc.
There is now a test for the _minimal desired outcome_ (search spans >1 node) which addresses the original user issue and looks to hold true over thousands of randomized test invocations.
As discussed, a repeatable test for the _optimal outcome_ (searches using all nodes) is not possible due to hash collisions so I implemented the "duel" regression testing approach you outlined where we repeat the hashing logic in the test to ensure there's at least agreement in the routing approach taken.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java:
```diff
@@ -367,6 +371,65 @@ public void testPreferNodes() throws InterruptedException, IOException {
+            Set<String> selectedNodes = new HashSet<>(numShards);
+            final GroupShardsIterator<ShardIterator> groupIterator = opRouting.searchShards(state, indexNames, null, sessionKey);
+
+            assertEquals("One group per index shard", numIndices * numShards, groupIterator.size());
```
nit: can you use assertThat or expose the actual values in the message.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java:
```diff
@@ -367,6 +371,65 @@ public void testPreferNodes() throws InterruptedException, IOException {
+                ShardRouting firstChoice = shardIterator.nextOrNull();
+                assertNotNull(firstChoice);
+                ShardRouting duelFirst = duelGetShards(state, firstChoice.shardId(), sessionKey).nextOrNull();
+                assertEquals("Regression test failure", duelFirst, firstChoice);
```
same comment about exposing values.

markharwood commented:
Rebasing before merge

ruria commented:
I think there are a "few" 2 data nodes clusters out there using Kibana. All we are using 50% of resources!! Rebalancing primary shards is the only way to go (I think) and it seems to me like a CRITICAL issue, maybe a backport is worth considering.

We ´ve tried to rebalance prymary shards acrros cluster using cluster.routing.rebalance.enable but it doesn´t work for us. I´m little confused here ( https://github.com/elastic/elasticsearch/issues/3293 ) Is it possible to rebalnace primary shards?

ruria commented:
Upss! Rebalancing primary shards not solves the issue, forget this. A user using Kibana (2 data node) is using just 50% of resources, and it have not workarround.

bleskes commented:
@ruria no, rebalancing the primaries wouldn't work. Do you only have one kibana users? multiple sessions should spread out the load.

ruria commented:
Yes, I have a few users. But for each of them, the "power" used is the half. Not all users are searching concurrently.

In our scenario, migrating to new hardware (x3 faster at indexing) from 2.3 to 5.4, we´re suffering a huge search time slowdown.
